src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -9,9 +10,34 @@ namespace GraphQL.Execution
     /// </summary>
     public static class ExecutionHelper
     {
+        private static readonly IDictionary<string, ArgumentValue> _emptyDirectiveArguments = new ReadOnlyDictionary<string, ArgumentValue>(new Dictionary<string, ArgumentValue>());
+
+        /// <summary>
+        /// Returns a dictionary of directives with their arguments values for a field.
+        /// Values will be retrieved from literals or variables as specified by the document.
+        /// </summary>
+        public static IDictionary<string, IDictionary<string, ArgumentValue>>? GetDirectives(GraphQLField field, Variables? variables, ISchema schema)
+        {
+            if (field.Directives == null || field.Directives.Count == 0)
+                return null;
+
+            var directives = new Dictionary<string, IDictionary<string, ArgumentValue>>(field.Directives.Count);
+
+            foreach (var dir in field.Directives.Items)
+            {
+                var dirDefinition = schema.Directives.Find(dir.Name);
+                if (dirDefinition == null)
+                    throw new ArgumentOutOfRangeException(nameof(field), $"Unknown directive '{dir.Name.StringValue}' for field '{field.Name.StringValue}'.");
```

Comment:
Question 2: wording of error message.

Reply:
I assume that validation rules should validate that directives are valid before executing the request?  If not, we should probably add that code.  Assuming that it does (or if we add a validation rule), maybe this would be better here:

```cs
if (dirDefinition == null)
    continue;
```

Then if someone purposely removed the validation rule, it would ignore unknown directives while executing the request.  I think that would be better.

Also, if by default the added validation rule is not part of the default set of validation rules, again I would think it would be better to skip unknown directives rather than throw an exception here.  For example:

```gql
{
  fieldWith2Literal @undefinedDirective @include(if: true)
  fieldWithoutDirectives @undefinedDirective
}
```

In the example above, executing `fieldWithoutDirectives` would not throw an exception assuming the field resolver does not attempt to read the directives at all, and `fieldWith2Literal` would always throw an exception, assuming the code was attempting to inspect the `include` directive.

So I think this code should ignore unknown directives and leave it to the validation rules to throw errors for unknown directives, if so desired.

Reply:
Agree, done.